apps/openmw/mwsound/soundmanagerimp.cpp

```diff
@@ -796,6 +804,96 @@ namespace MWSound
         }
     }
 
+    void SoundManager::updateWaterSound(float /*duration*/)
+    {
+        MWBase::World* world = MWBase::Environment::get().getWorld();
+        const MWWorld::ConstPtr player = world->getPlayerPtr();
+        osg::Vec3f pos = player.getRefData().getPosition().asVec3();
+
+        float volume = 0.0f;
+        std::string soundId = player.getCell()->isExterior() ? mNearWaterOutdoorID : mNearWaterIndoorID;
+
+        if (!mListenerUnderwater)
+        {
+            if (player.getCell()->getCell()->hasWater())
+            {
+                float dist = std::abs(player.getCell()->getWaterLevel() - pos.z());
+
+                if (player.getCell()->isExterior() && dist < mNearWaterOutdoorTolerance)
+                {
+                    volume = (mNearWaterOutdoorTolerance - dist) / mNearWaterOutdoorTolerance;
+
+                    int underwaterPoints = 0;
+
+                    if (mNearWaterPoints > 1)
+                    {
+                        float step = mNearWaterRadius * 2.0f / (mNearWaterPoints - 1);
+
+                        for (int x = 0; x < mNearWaterPoints; x++)
+                        {
+                            for (int y = 0; y < mNearWaterPoints; y++)
+                            {
+                                float height = world->getTerrainHeightAt(
+                                            osg::Vec3f(pos.x() - mNearWaterRadius + x*step, pos.y() - mNearWaterRadius + y*step, 0.0f));
+
+                                if (height < 0)
+                                    underwaterPoints++;
+                            }
+                        }
+                    }
+
+                    volume *= underwaterPoints * 2.0f / (mNearWaterPoints*mNearWaterPoints);
```

Comment:
This looks weird. The maximum possible volume depends on ```mNearWaterPoints```, such that when ```mNearWaterPoints == 10``` this can't be any higher than ```0.2```, and when ```mNearWaterPoints == 20``` it can't be any higher than ```0.1```. Perhaps ```(underwaterPoints*underwaterPoints) / (mNearWaterPoints*mNearWaterPoints)``` would work better?